File: core/src/main/java/org/elkd/core/consensus/messages/Entry.java
package org.elkd.core.consensus.messages;

import com.google.common.base.Preconditions;

import javax.annotation.Nonnull;
import java.util.Objects;

public class Entry {
  public static final Entry NULL_ENTRY = Entry.builder(0, "default").build();

  private final int mTerm;
  private final String mEvent;

  private Entry(final int term, @Nonnull final String event) {
    mTerm = term;
    mEvent = Preconditions.checkNotNull(event, "event");
  }

  public static Builder builder(final int term, final String event) {
    return new Builder(term, event);
  }

  public static class Builder {
    private int mTerm;
    private String mEvent;

    Builder(final int term, @Nonnull final String event) {
      mTerm = term;
      mEvent = Preconditions.checkNotNull(event, "event");
    }

    public Entry build() {
      return new Entry(
          mTerm,
          mEvent
      );
    }
  }

  public int getTerm() {
    return mTerm;
  }

  public String getEvent() {
    return mEvent;
  }

  @Override
  public boolean equals(final Object rhs) {
    if (this == rhs) {
      return true;
    }
    if (rhs == null || getClass() != rhs.getClass()) {
      return false;
    }
    final Entry entry = (Entry) rhs;
    return mTerm == entry.mTerm &&
        Objects.equals(mEvent, entry.mEvent);
  }

  @Override
  public int hashCode() {
    return Objects.hash(mTerm, mEvent);
  }

  @Override
  public String toString() {
    return "Entry{" +
        "mTerm=" + mTerm +
        ", mEvent='" + mEvent + '\'' +
        '}';
  }
}

File: core/src/main/java/org/elkd/core/consensus/messages/Entry.kt
package org.elkd.core.consensus.messages

import java.util.UUID
import org.elkd.core.log.LogEntry

class Entry private constructor(
    override val term: Int,
    override val uuid: String,
    val kvs: List<KV>
) : LogEntry {
  override fun toString(): String {
    return "Entry(uuid=$uuid, term=$term, kvs=$kvs)"
  }

  companion object {
    @Deprecated("Logs will start from indexing from zero")
    @JvmStatic val NULL_ENTRY = builder(0, "null").build()

    @JvmStatic fun builder(term: Int, uuid: String = UUID.randomUUID().toString()) = Builder(term, uuid)
  }

  class Builder internal constructor(
      val term: Int,
      val uuid: String,
      private val kvs: MutableList<KV> = mutableListOf()
  ) {
    fun addKV(kv: KV): Builder {
      kvs.add(kv)
      return this
    }

    fun addAllKV(kvs: List<KV>): Builder {
      this.kvs.addAll(kvs)
      return this
    }

    fun build() = Entry(term, uuid, kvs)
  }
}
